File: TicTacToe_MiniMax/gato.py
import os


def imprimeTablero(tablero):
    print(tablero[1] + '|' + tablero[2] + '|' + tablero[3])
    print('-----')
    print(tablero[4] + '|' + tablero[5] + '|' + tablero[6])
    print('-----')
    print(tablero[7] + '|' + tablero[8] + '|' + tablero[9])
    print("\n")

def existeEspacio(posicion):
    if posicion > 9:
        return False
    if tablero[posicion] == ' ':
        return True
    else:
        return False

def insertaFicha(ficha, posicion):
    if existeEspacio(posicion):
        tablero[posicion] = ficha
        imprimeTablero(tablero)
        if (checaEmpate()):
            print("[i]Empate")
            exit()
        if checaGana():
            if ficha == 'X':
                print("[i]Gana la IA")
                exit()
            else:
                print("[i]Gana el jugador")
                exit()
        return
    else:
        print("[!] No se puede insertar la ficha en {1}".format(posicion))
        posicion = int(input("[+] Ingresa un nueva casilla \n-> "))
        insertaFicha(ficha, posicion)
        return

def checaGana():
    if (tablero[1] == tablero[2] and tablero[1] == tablero[3] and tablero[1] != ' '):
        return True
    elif (tablero[4] == tablero[5] and tablero[4] == tablero[6] and tablero[4] != ' '):
        return True
    elif (tablero[7] == tablero[8] and tablero[7] == tablero[9] and tablero[7] != ' '):
        return True
    elif (tablero[1] == tablero[4] and tablero[1] == tablero[7] and tablero[1] != ' '):
        return True
    elif (tablero[2] == tablero[5] and tablero[2] == tablero[8] and tablero[2] != ' '):
        return True
    elif (tablero[3] == tablero[6] and tablero[3] == tablero[9] and tablero[3] != ' '):
        return True
    elif (tablero[1] == tablero[5] and tablero[1] == tablero[9] and tablero[1] != ' '):
        return True
    elif (tablero[7] == tablero[5] and tablero[7] == tablero[3] and tablero[7] != ' '):
        return True
    else:
        return False

def checaGanaFicha(mark):
    if tablero[1] == tablero[2] and tablero[1] == tablero[3] and tablero[1] == mark:
        return True
    elif (tablero[4] == tablero[5] and tablero[4] == tablero[6] and tablero[4] == mark):
        return True
    elif (tablero[7] == tablero[8] and tablero[7] == tablero[9] and tablero[7] == mark):
        return True
    elif (tablero[1] == tablero[4] and tablero[1] == tablero[7] and tablero[1] == mark):
        return True
    elif (tablero[2] == tablero[5] and tablero[2] == tablero[8] and tablero[2] == mark):
        return True
    elif (tablero[3] == tablero[6] and tablero[3] == tablero[9] and tablero[3] == mark):
        return True
    elif (tablero[1] == tablero[5] and tablero[1] == tablero[9] and tablero[1] == mark):
        return True
    elif (tablero[7] == tablero[5] and tablero[7] == tablero[3] and tablero[7] == mark):
        return True
    else:
        return False

def checaEmpate():
    for key in tablero.keys():
        if (tablero[key] == ' '):
            return False
    return True

def movimientoJugador():
    posicion = int(input("[+] Ingresa una casilla para tirar 'O'\n-> "))
    insertaFicha(jugadorFicha, posicion)
    return

def movimientoComputadora():
    bestScore = -800
    bestMove = 0
    for key in tablero.keys():
        # checa el peso de cada lugar
        if (tablero[key] == ' '):
            tablero[key] = computadoraIA
            # Optiene un score
            score = minimax(tablero, 0, False)
            tablero[key] = ' '
            # optiene el mejor resultado
            if (score > bestScore):
                bestScore = score
                bestMove = key

    insertaFicha(computadoraIA, bestMove)
    return

def minimax(tablero, depth, isMaximizing):
    if (checaGanaFicha(computadoraIA)):
        return 1
    elif (checaGanaFicha(jugadorFicha)):
        return -1
    elif (checaEmpate()):
        return 0

    if (isMaximizing): # Es maximo
        bestScore = -800
        for key in tablero.keys():
            if (tablero[key] == ' '):
                tablero[key] = computadoraIA
                score = minimax(tablero, depth + 1, False)
                tablero[key] = ' '
                if (score > bestScore):
                    bestScore = score
        return bestScore
    else: # es minimo
        bestScore = 800
        for key in tablero.keys():
            if (tablero[key] == ' '):
                tablero[key] = jugadorFicha
                score = minimax(tablero, depth + 1, True)
                tablero[key] = ' '
                if (score < bestScore):
                    bestScore = score
        return bestScore

def clearScreen():
    if os.name == 'posix':
        _ = os.system('clear')
    else:
        _ = os.system('cls')

def casillasDisponibles():
    print("[i] Estas son las casillas:")
    print("1, 2, 3 ")
    print("4, 5, 6 ")
    print("7, 8, 9 \n")


tablero = {1: ' ', 2: ' ', 3: ' ',
        4: ' ', 5: ' ', 6: ' ',
        7: ' ', 8: ' ', 9: ' '}

imprimeTablero(tablero)

while True:
    clearScreen()
    turno = int(input("[+] Eliga \"1\" para ser el primer jugador o \"2\" para ser el segundo\n-> "))
    if turno == 1 or turno == 2 :
        break

# 
jugadorFicha = 'O'

computadoraIA = 'X'

while not checaGana():
    print("-------------------------------")
    casillasDisponibles()
    if turno == 1:
        movimientoJugador()
        movimientoComputadora()
    else:
        movimientoComputadora()
        movimientoJugador()

File: TicTacToe_MiniMax/gato.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"syscall"
)

func cleanScreen() {
	c := exec.Command("clear")
	c.Stdout = os.Stdout
	c.Run()
}

func imprimeTablero(tablero map[uint]string) {
	fmt.Println(tablero[1] + "|" + tablero[2] + "|" + tablero[3])
	fmt.Println("-----")
	fmt.Println(tablero[4] + "|" + tablero[5] + "|" + tablero[6])
	fmt.Println("-----")
	fmt.Println(tablero[7] + "|" + tablero[8] + "|" + tablero[9])
	fmt.Print("\n")
}

func casillasDisponibles() {
	fmt.Println("[i] Estas son las casillas:")
	fmt.Println("1, 2, 3 ")
	fmt.Println("4, 5, 6 ")
	fmt.Println("7, 8, 9 ")
}

func existeEspacio(tablero map[uint]string, posicion uint) bool {
	if posicion > 9 {
		return false
	}
	if tablero[posicion] == "-" {
		return true
	}
	return false
}

func checaEmpate(tablero map[uint]string) bool {
	for _, v := range tablero {
		if v == "-" {
			return false
		}
	}
	return true
}

func checaGana(tablero map[uint]string) bool {
	if tablero[1] == tablero[2] && tablero[1] == tablero[3] && tablero[1] != "-" {
		return true
	}
	if tablero[4] == tablero[5] && tablero[4] == tablero[6] && tablero[4] != "-" {
		return true
	}
	if tablero[7] == tablero[8] && tablero[7] == tablero[9] && tablero[7] != "-" {
		return true
	}
	if tablero[1] == tablero[4] && tablero[1] == tablero[7] && tablero[1] != "-" {
		return true
	}
	if tablero[2] == tablero[5] && tablero[2] == tablero[8] && tablero[2] != "-" {
		return true
	}
	if tablero[3] == tablero[6] && tablero[3] == tablero[9] && tablero[3] != "-" {
		return true
	}
	if tablero[1] == tablero[5] && tablero[1] == tablero[9] && tablero[1] != "-" {
		return true
	}
	if tablero[7] == tablero[5] && tablero[7] == tablero[3] && tablero[7] != "-" {
		return true
	}
	return false
}

func checaGanaFicha(tablero map[uint]string, mark string) bool {
	if tablero[1] == tablero[2] && tablero[1] == tablero[3] && tablero[1] == mark {
		return true
	}
	if tablero[4] == tablero[5] && tablero[4] == tablero[6] && tablero[4] == mark {
		return true
	}
	if tablero[7] == tablero[8] && tablero[7] == tablero[9] && tablero[7] == mark {
		return true
	}
	if tablero[1] == tablero[4] && tablero[1] == tablero[7] && tablero[1] == mark {
		return true
	}
	if tablero[2] == tablero[5] && tablero[2] == tablero[8] && tablero[2] == mark {
		return true
	}
	if tablero[3] == tablero[6] && tablero[3] == tablero[9] && tablero[3] == mark {
		return true
	}
	if tablero[1] == tablero[5] && tablero[1] == tablero[9] && tablero[1] == mark {
		return true
	}
	if tablero[7] == tablero[5] && tablero[7] == tablero[3] && tablero[7] == mark {
		return true
	}
	return false
}

func insertaFicha(tablero map[uint]string, ficha string, posicion uint) {
	if existeEspacio(tablero, posicion) {
		tablero[posicion] = ficha
		imprimeTablero(tablero)
		if checaEmpate(tablero) {
			fmt.Println("[i] Empate")
			syscall.Exit(0)
		}
		if checaGana(tablero) {
			if ficha == "X" {
				fmt.Println("[i] Gana la IA")
				syscall.Exit(0)
			} else {
				fmt.Println("[i] Gana el jugador")
				syscall.Exit(0)
			}
		}
	} else {
		cleanScreen()
		casillasDisponibles()
		imprimeTablero(tablero)
		fmt.Printf("[!] No se puede insertar la ficha en %d\n", posicion)
		fmt.Print("[+] Ingresa un nueva casilla \n-> ")
		fmt.Scanf("%d", &posicion)
		insertaFicha(tablero, ficha, posicion)
	}
}

func movimientoJugador(tablero map[uint]string, ficha string) {
	var posicion uint
	fmt.Print("[+] Ingresa una casilla para tirar 'O'\n-> ")
	fmt.Scanf("%d", &posicion)
	insertaFicha(tablero, "O", posicion)
}

func movimientoComputadora(tablero map[uint]string, ficha string) {
	var bestMove uint
	bestScore := -800
	bestMove = 0
	for key, ficha := range tablero {
		if ficha == "-" {
			tablero[key] = ficha
			score := minimax(tablero, 0, false)
			tablero[key] = "-"
			if score > bestScore {
				bestScore = score
				bestMove = key
			}
		}
	}
	insertaFicha(tablero, ficha, bestMove)
}

func minimax(tablero map[uint]string, depth int, isMaximizing bool) int {
	if checaGanaFicha(tablero, "X") {
		return 1
	}
	if checaGanaFicha(tablero, "O") {
		return -1
	}
	if checaEmpate(tablero) {
		return 0
	}

	if isMaximizing {
		bestScore := -800
		for key, ficha := range tablero {
			if ficha == "-" {
				tablero[key] = "X"
				score := minimax(tablero, depth+1, false)
				tablero[key] = "-"
				if score > bestScore {
					bestScore = score
				}
			}
		}
		return bestScore
	} else {
		bestScore := 800
		for key, ficha := range tablero {
			if ficha == "-" {
				tablero[key] = "O"
				score := minimax(tablero, depth+1, true)
				tablero[key] = "-"
				if score < bestScore {
					bestScore = score
				}

			}
		}
		return bestScore
	}
}

func main() {
	tablero := map[uint]string{1: "-", 2: "-", 3: "-",
		4: "-", 5: "-", 6: "-",
		7: "-", 8: "-", 9: "-"}

	fmt.Println("\t[i] Juego del Gato MiniMax en Golang")
	turno := 0
	for {
		fmt.Print("[+] Eliga \"1\" para ser el primer jugador o \"2\" para ser el segundo\n-> ")
		fmt.Scanf("%d", &turno)
		if turno == 1 || turno == 2 {
			break
		}
		cleanScreen()
	}
	for {
		if checaGana(tablero) {
			break
		}
		fmt.Println("-------------------------------")
		casillasDisponibles()
		if turno == 1 {
			movimientoJugador(tablero, "O")
			movimientoComputadora(tablero, "X")
		} else {
			movimientoComputadora(tablero, "X")
			movimientoJugador(tablero, "O")
		}
	}
}
